### ISSUE TITLE
Calibration features not saved in model state_dict after training (QAT)

### ISSUE THREAD

**Issue by lorsonblair** [foundation-model-stack/fms-model-optimizer#98]
I wasn't sure which issue this fell under, so I created a blank issue. I'm working with fms_mo for quantization. Our platform has maximum run times so I use checkpointing to resume QAT when server runtime times out (expires). However, I noticed that calibration features, which are added to the model during qmodel_prep, are not saved in the model state_dict after training. This causes a load error when resuming training from a checkpoint. The main reason for this is as mentioned prior, the new calib features added  during qmodel prep are not saved. Two examples of these features are: 
 
"model.backbone.encoder.layers.0.self_attn.sdp_attn.QBmm376.quantize_calib_m2.cv_movAvg" and 
"model.backbone.encoder.layers.0.self_attn.sdp_attn.QBmm376.quantize_calib_m2.zero_point"

I'm required to run qmodel_prep when resuming, to get a quantized model that matches the saved model. Is there a workaround for this? Loading with strict = False leads to nan during inference. Let me know if you require additional information like code snippet etc. Thanks in advance.

**Comment by chichun-charlie-liu** [comment]
Thanks for your input. We didn't test this type of set-up, so it's not well considered before. We'll take a closer look and prepare a fix. (We are improving the ckpt saving function at the moment, which would be useful to address this issue.) 

a code snippet would be helpful for the debugging, as this seems to be an sdpa case but somehow QBmm still got attached. Usually they are mutual exclusive... so it would be nice to see an example to reproduce the error for our debugging.

Meanwhile, a possible work-around would be: 
In your second/resume run, you can set the `qmodel_calibration` and `qmodel_calibration_new` to 0, because once you run `qmodel_prep` and load the ckpt, we should not need to trigger/use those in `quantize_calib_mx`, we should have the "real" quant params stored in `quantize_mx` already. The calib ones are for initial guess only in the QAT cases.

Let us know if this helps. Thanks.

**Comment by lorsonblair** [comment]
Thanks for your response. So, I've tried several workarounds, including your suggestions. I incorporated `qmodel_calibration_new = 0` after you suggested it. The combinations and errors produced are summarized below: 

1. `qmodel_calibration = 0` and `qmodel_calibration_new = 0` - Missing and unexpected keys in state_dict
2. `qmodel_calibration = n` and `qmodel_calibration_new = 0` - Missing keys in state_dict
3. `qmodel_calibration = 0` and `qmodel_calibration_new = n` - Missing keys in state_dict
4. `qmodel_calibration = n` and `qmodel_calibration_new = n` - Missing keys in state_dict 

`qmodel_calibration_new = n` was responsible for the majority of the missing keys, including keys with `calib_feature`, and `calib_m1_...` and `calib_m2_...` which are part of the QBmm features. The configuration `qmodel_calibration = n` and `qmodel_calibration_new = 0` produced the least number of missing keys. Only the keys with `calib_counter` were missing. I observed that `qmodel_calibration = n` is needed for `obsv` features, since they show up as "Unexpected keys" in state_dict when `qmodel_calibration = 0`. 

Update: 

Setting `qmodel_calibration = 0` and `qmodel_calibration_new = 0` on resume and setting `strict=False` when loading the saved checkpoint state_dict worked, including getting none nan metric values for testing. Before, having `qmodel_calibration_new = n` and `strict = False` resumed training, however, testing resulted in nan values (for the performance metrics: mse, mae, rmse). I believe testing failed with this configuration because the `calib_feature` keys loading during calibration triggered by `qmodel_calibration_new = n` became instrumental in the model and not having them loaded during checkpoint resume and testing became an issue. 

Below are some code snippet of saving and loading

Preparing model for quantization
`qmodel_prep(self.model, data_mb, self.qcfg, optimizer=optimizer, save_fname=f"{base_name}_graph.pt",                                                            qlast=self.args.qlast)`

Saving a model:

`def save(self, model, path):
      state = {}
      self.save_to_state(model, state)

 path.parent.mkdir(parents=True, exist_ok=True)
 torch.save(state, path)`

 def save_to_state(self, model, state):
      state["model_hash"]  = self.model_hash
      state["model_state"] = model.state_dict()`

Loading saved checkpoint:

`def _default_dict_loader(model, model_state):
       model.load_state_dict(model_state)
       # strict = False loads model from checkpoint. Testing works if `qmodel_calibration = 0` and `qmodel_calibration_new = 
       0

 def load(self, model, path, device, checkHash=True):

      state = torch.load(path, map_location=device)
      self.load_from_state(model, state, checkHash)

 def load_from_state(self, model, state, checkHash=True):
      if checkHash and self.model_hash != state["model_hash"]:
           msg = "ERROR: saved model does not match current model params"
           print(msg) # log if we are logging
           print(msg, file=sys.stderr)
           exit(1)

    self.dict_loader(model, state["model_state"])`

Let me know if you want me to post the all the Missing and Unexpected keys from the error messages for the different calibration combinations (configs).

**Comment by chichun-charlie-liu** [comment]
just want to emphasize that when doing resume/testing from a QAT process, we really don't want to use "calibration", because the trained quantization parameters such as scales and clip_vals will be reset by the calibration. And a great portion of the training efforts will be lost. Calibration is only needed when we kick off the training the very first time, because we want to have a reasonable initial guess of scales, clip_vals. 
Seems like you can get the resume working now. Feel free to let us know if the testing issue persists.

**Comment by lorsonblair** [comment]
Yes, I suspected that. That's why I first tested with setting `qmodel_calibration = 0`. I had forgotten about `qmodel_calibration_new` also had to be set to 0 until you mentioned it. Testing is working fine now also. Thanks again for your help. 

Also, could you explain why `qmodel_prep` doesn't return the model or optimizer, but it returns model if no optimizer is provided as an argument?

**Comment by chichun-charlie-liu** [comment]
apparently it was a misaligned indentation... 8) (will fix it in the next related PR)
but historically speaking, when models were small and we could afford to have a copy of the original, returning a `qmodel` while keeping the original unchanged would make sense. But now we prefer not to for obvious reason. and since we already make changes to the model inplace, returning a model would not add much...
